pkg/instructor/chat.go

```diff
@@ -48,12 +62,66 @@ func chatHandler(i Instructor, ctx context.Context, request interface{}, respons
 			if err != nil {
 				// TODO:
 				// add more sophisticated retry logic (send back validator error and parse error for model to fix).
+
+				countUsageFromResp(resp, usage)
 				continue
 			}
 		}
 
-		return resp, nil
+		return addUsageSumToResp(resp, usage), nil
 	}
 
-	return nil, errors.New("hit max retry attempts")
+	return i.EmptyResponseWithUsage(usage), errors.New("hit max retry attempts")
+}
+
+func nilChatRespWithUsage(response interface{}) interface{} {
```

Comment:
we should add the usage method as an abstract method to be implemented in each provider, so we do not need to worry about provider specific logic in this file